stablehlo/tests/CheckOps.cpp

```diff
@@ -156,6 +157,96 @@ llvm::Error evalExpectSerializedEqOp(const Tensor &expected, StringRef probeId,
   return evalExpectEqOp(expected, *tensor);
 }
 
+static uint64_t ULPDifference(APFloat f, APFloat g) {
+  // z is ULP-distance between exact 0 and largest subnormal:
+  uint64_t z = APFloat::getSmallestNormalized(f.getSemantics())
+                   .bitcastToAPInt()
+                   .getLimitedValue() -
+               1;
+  // a is ULP-distance between exact 0 and abs(f):
+  uint64_t a = (f.isNegative() ? -f : f).bitcastToAPInt().getLimitedValue();
+  // b is ULP-distance between exact 0 and abs(g):
+  uint64_t b = (g.isNegative() ? -g : g).bitcastToAPInt().getLimitedValue();
+  if (f.isFinite() && g.isFinite()) {
+    // subtract subnormals contribution, round subnormals to closest normal or
+    // zero:
```

Comment:
Why round subnormals to the closest normal or 0? The ULP difference between 0 and the smallest normal number should be the number of subnormal values, not 1, right?

Reply:
> Why round subnormals to the closest normal or 0?
> The ULP difference between 0 and the smallest normal number should be the number of subnormal values, not 1, right?

Correct. My suggestion is to flush subnormals to closest normal value so that testing closeness of math functions values would be device as well as platform agnostic (some devices/platforms have FTZ mode enabled by default).

Hmm, let me check the consequences of accounting subnormals when testing math functions and get back here, perhaps accounting subnormals ought to be optional...

Reply:
Update: it turns out that the generated implementations work with subnormals just fine. I'll remove subnormals flushing..

Reply:
Rounding subnormals is now removed.